experiments/download_ssl4eo.py

```diff
@@ -0,0 +1,738 @@
+from shapely.geometry import Point, shape
+from torchvision.datasets.utils import download_and_extract_archive
+from tqdm import tqdm
+
+warnings.simplefilter("ignore", UserWarning)
+
+ALL_BANDS_L2A = [
+    "B1",
+    "B2",
+    "B3",
+    "B4",
+    "B5",
+    "B6",
+    "B7",
+    "B8",
+    "B8A",
+    "B9",
+    "B11",
+    "B12",
+]
+ALL_BANDS_L1C = [
+    "B1",
+    "B2",
+    "B3",
+    "B4",
+    "B5",
+    "B6",
+    "B7",
+    "B8",
+    "B8A",
+    "B9",
+    "B10",
+    "B11",
+    "B12",
+]
+RGB_BANDS = ["B4", "B3", "B2"]
+ALL_BANDS_GRD = ["VV", "VH"]
+
+
+""" samplers to get locations of interest points"""
+
+
+class UniformSampler:
+    def sample_point(self) -> List[float]:
+        lon = np.random.uniform(-180, 180)
+        lat = np.random.uniform(-90, 90)
+        return [lon, lat]
+
+
+class GaussianSampler:
+    def __init__(
+        self,
+        interest_points: Optional[List[List[float]]] = None,
+        num_cities: int = 1000,
+        std: float = 20,
+    ) -> None:
+        if interest_points is None:
+            cities = self.get_world_cities()
+            self.interest_points = self.get_interest_points(cities, size=num_cities)
+        else:
+            self.interest_points = interest_points
+        self.std = std
+
+    def sample_point(self) -> List[float]:
+        rng = np.random.default_rng()
+        point = rng.choice(self.interest_points)
+        std = self.km2deg(self.std)
+        lon, lat = np.random.normal(loc=point, scale=[std, std])
+        return [lon, lat]
+
+    @staticmethod
+    def get_world_cities(
+        download_root: str = os.path.expanduser("./world_cities/"),
+    ) -> List[Dict[str, Any]]:
+        url = "https://simplemaps.com/static/data/world-cities/basic/simplemaps_worldcities_basicv1.71.zip"  # noqa
+        filename = "worldcities.csv"
+        if not os.path.exists(os.path.join(download_root, os.path.basename(url))):
+            download_and_extract_archive(url, download_root)
+        with open(os.path.join(download_root, filename), encoding="UTF-8") as csvfile:
+            reader = csv.DictReader(csvfile, delimiter=",", quotechar='"')
+            cities = []
+            for row in reader:
+                row["population"] = (
+                    row["population"].replace(".", "") if row["population"] else "0"
+                )
+                cities.append(row)
+        return cities
+
+    @staticmethod
+    def get_interest_points(
+        cities: List[Dict[str, str]], size: int = 10000
+    ) -> List[List[float]]:
+        cities = sorted(cities, key=lambda c: int(c["population"]), reverse=True)[:size]
+        points = [[float(c["lng"]), float(c["lat"])] for c in cities]
+        return points
+
+    @staticmethod
+    def km2deg(kms: float, radius: float = 6371) -> float:
+        return kms / (2.0 * radius * np.pi / 360.0)
+
+    @staticmethod
+    def deg2km(deg: float, radius: float = 6371) -> float:
+        return deg * (2.0 * radius * np.pi / 360.0)
```

Comment:
Aren't these only true at the equator? We might want to fix this in a follow-up PR.

Reply:
Yep, this was extremely simplified.

Reply:
Actually, now that I think about it, the patches we download are in WGS 84 (lat, long), so the conversion is correct if we want to ensure that all patches have the same size.

Is there a reason WGS 84 was chosen? Is that the default projection of the data stored in GEE? It will result in a lot of distortion in higher latitudes, although there aren't many cities at higher latitudes.

Reply:
GEE does have a default projection WGS 84, but you can export to other projections. The main reason is still to make it simple, as the distortion can be kind of neglected (considering the pixel resolution).